server/lib/routes/secret.js

```diff
@@ -0,0 +1,72 @@
+'use strict'
+
+async function deleteTokenIfAuthorized(
+  firebaseAdmin,
+  tokenId,
+  subscriptionId,
+  user
+) {
+  const db = firebaseAdmin.firestore()
+  const subscriptionRef = await db
+    .collection('subscriptions')
+    .where(firebaseAdmin.firestore.FieldPath.documentId(), '==', subscriptionId)
+    .where('userId', '==', user)
+    .get()
+
+  if (subscriptionRef.empty) {
+    return false
+  }
+
+  await db
+    .collection('allTokens')
+    .doc(tokenId)
+    .delete()
+
+  return true
+}
+
+async function secretRoutes(server) {
+  server.route({
+    method: 'DELETE',
+    url: '/api/secret/:secretId',
+    preHandler: server.auth([server.authenticate]),
+    handler: async (request, reply) => {
+      const { firebaseAdmin } = server
+      const { secretId } = request.params
+
+      const db = firebaseAdmin.firestore()
+
+      const tokensForSecret = await db
+        .collection('allTokens')
+        .where('secretId', '==', secretId)
+        .get()
+
+      if (tokensForSecret.empty) {
+        return reply.notFound('Secret not found')
+      }
+
+      let deletedError = false
+
+      for await (const didDelete of await tokensForSecret.docs.map((doc) =>
+        deleteTokenIfAuthorized(
+          firebaseAdmin,
+          doc.id,
+          doc.data().subscriptionId,
+          request.user
+        )
+      )) {
+        if (!didDelete) {
+          deletedError = true
+        }
+      }
```

Comment:
doesn't firestore support a single query to delete multiple records?

Reply:
The reason I went this route is because each token stores the subscription id so I thought the simplest way was to just iterate over all of these tokens and do the check in `deleteTokenIfAuthorized` to make sure the user can delete that particular item. I do realise that the subscription id is going to be the same for each token associated with a secret so the data is a little duplicated leading to solution I came to here.